crypto/evp/keymgmt_lib.c

```diff
@@ -211,3 +225,260 @@ void *evp_keymgmt_util_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
+    void *tmp_keydata;
+
+    tmp_keydata = evp_keymgmt_util_export_to_provider(pk1, tmp_keymgmt2);
+
+    if (tmp_keydata != NULL) {
+        tmp_keymgmt1 = tmp_keymgmt2;
+        tmp_keydata1 = tmp_keydata;
+    } else {
+        tmp_keydata = evp_keymgmt_util_export_to_provider(pk2, tmp_keymgmt1);
+
+        if (tmp_keydata != NULL) {
+            tmp_keymgmt2 = tmp_keymgmt1;
+            tmp_keydata2 = tmp_keydata;
+        }
+    }
+
+    *keymgmt1 = tmp_keymgmt1;
+    *keymgmt2 = tmp_keymgmt2;
+    *keydata1 = tmp_keydata1;
+    *keydata2 = tmp_keydata2;
+}
+
+
+static int match_type(EVP_KEYMGMT *keymgmt1, EVP_KEYMGMT *keymgmt2)
+{
+    /* Unconstify the provider, because evp_first_name() demands it */
+    OSSL_PROVIDER *prov2 = (OSSL_PROVIDER *)EVP_KEYMGMT_provider(keymgmt2);
+    const char *name2 = evp_first_name(prov2, EVP_KEYMGMT_number(keymgmt2));
+
+    return EVP_KEYMGMT_is_a(keymgmt1, name2);
+}
+
+/*
+ * evp_keymgmt_util_match() adheres to the return values that EVP_PKEY_cmp()
+ * and EVP_PKEY_cmp_parameters() return, i.e.:
+ *
+ *  1   same key
+ *  0   not same key
+ * -1   not same key type
+ * -2   unsupported operation
+ */
+static int implements_match(EVP_KEYMGMT *keymgmt)
+{
+    return keymgmt->match != NULL;
+}
+
+int evp_keymgmt_util_match(EVP_PKEY *pk1, EVP_PKEY *pk2, int selection)
+{
+    EVP_KEYMGMT *impmatch_keymgmt1 = NULL, *impmatch_keymgmt2 = NULL;
+    void *keydata1 = NULL, *keydata2 = NULL;
+
+    find_keymgmt(&impmatch_keymgmt1, &impmatch_keymgmt2, &keydata1, &keydata2,
+                 pk1, pk2, &implements_match);
+
+    /* If we found no suitable keymgmt for either key, support is missing */
+    if (impmatch_keymgmt1 == NULL && impmatch_keymgmt2 == NULL)
+        return -2;
+
+    /*
+     * If we don't have matching keymgmt implementations, we check that they
+     * handle the same key type.
+     * We trust that aliases are properly registered.
+     */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2) {
+        if (impmatch_keymgmt1 == NULL
+            || impmatch_keymgmt2 == NULL
+            || !match_type(impmatch_keymgmt1, impmatch_keymgmt2)) {
+            ERR_raise(ERR_LIB_EVP, EVP_R_DIFFERENT_KEY_TYPES);
+            return -1;           /* Not the same type */
+        }
+    }
+
+    /*
+     * If the two keymgmt aren't the same, try to prepare any of the two
+     * EVP_PKEYs so they end up matching.
+     */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2)
+        prepare_binary_op(&impmatch_keymgmt1, &impmatch_keymgmt2,
+                          &keydata1, &keydata2, pk1, pk2);
+
+    /* If we still don't have matching keymgmt implementations, we give up */
+    if (impmatch_keymgmt1 != impmatch_keymgmt2)
+        return -2;
+
+    return evp_keymgmt_match(impmatch_keymgmt1, keydata1, keydata2, selection);
+}
+
+static int implements_copy_and_export(EVP_KEYMGMT *keymgmt)
+{
+    return keymgmt->copy != NULL
+        && (keymgmt->import == NULL || keymgmt->export != NULL);
```

Comment:
Are you sure you mean  keymgmt->import == NULL ?
Below it says 'only implement |copy|, but must also implement |export| if it
     * implements |import|'

Reply:
> but must also implement |export| if it implements |import|

That translates into: "it's ok if |import| isn't implemented, then we don't care about checking for |export|.  However, if |import| *is* implemented, then so must |export|"

In logical terms:
1.  IF |import| == NULL THEN True
2.  OTHERWISE, IF |export| != NULL THEN True
3.  OTHERWISE False

Makes sense?

Reply:
I don't understand.  A provider can do import-only or export-only or both.  Is that concept not supported by this change?

Reply:
Hmmm, good question.  I'll have a look again.